src/light_data/light_store.rs

```diff
@@ -132,36 +136,109 @@ impl<F: LurkField> LightStore<F> {
         Ok(path)
     }
 
+    fn intern_ptr_data(
+        &self,
+        ptr0: ScalarPtr<F>,
+        store: &mut ScalarStore<F>,
+    ) -> anyhow::Result<()> {
+        let mut stack = Vec::new();
+        stack.push(ptr0);
+        while let Some(ptr) = stack.pop() {
+            if store.get_expr(&ptr).is_none() {
+                if ptr.is_immediate() {
```

Comment:
It would be great if you could break this up in two handler functions. Something like this perhaps?
```rust
fn handle_immediate(tag: ExprTag, value: Value) -> anyhow::Result<ScalarExpression> { ... }

fn handle_non_immediate(
    tag: ExprTag,
    ptr: ScalarPtr<F>,
    expr: &LightExpr,
    store: &mut ScalarStore<F>,
) -> anyhow::Result<ScalarExpression> 
```